File: src/state/recover.ts
import { INITIAL_STATE, TRANSITIONS, type State } from "./model";
import type { LogEntry, TransitionLog } from "./log";
import { checksumOf } from "./log";

// #EXT-002-REQ-4 Start
/**
 * Crash recovery by log replay ([REQ-4]). After a process or node crash, the
 * current state is reconstructed deterministically by replaying the durable log
 * entries `1..N` in order. Replay is total and deterministic: the same log
 * always yields the same state.
 *
 * Torn / interrupted final commit: `TransitionLog.read()` already drops a
 * partially-written trailing line (a line with no terminating newline parses to
 * `null`). Recovery additionally validates each entry's `index` continuity and
 * `checksum`; an entry that is corrupt or out of sequence is treated as the torn
 * boundary and replay stops there, discarding it and anything after. The result
 * is the consistent state that existed immediately before the crash.
 */

/** Outcome of a recovery: the reconstructed state and how many entries applied. */
export interface RecoveryResult {
  /** The deterministically reconstructed current state. */
  readonly state: State;
  /** Number of log entries successfully replayed (the durable prefix). */
  readonly appliedCount: number;
  /** True if a torn/corrupt trailing entry was discarded during replay. */
  readonly discardedTornEntry: boolean;
}

/**
 * True when `entry` is internally consistent at the expected 1-indexed position:
 * its `index` matches, its `(from,event,to)` is a declared transition, and its
 * `checksum` matches the recomputed value. A failing entry marks the torn
 * boundary.
 */
function isEntryConsistent(entry: LogEntry, expectedIndex: number): boolean {
  if (entry.index !== expectedIndex) {
    return false;
  }
  const declaredTo = TRANSITIONS[entry.from]?.[entry.event];
  if (declaredTo === undefined || declaredTo !== entry.to) {
    return false;
  }
  return entry.checksum === checksumOf(entry.index, entry.from, entry.event, entry.to);
}

/**
 * Replay the durable log to reconstruct current state. Starts from
 * {@link INITIAL_STATE} and applies each consistent entry in order; the first
 * inconsistent (torn/corrupt/out-of-sequence) entry halts replay and is
 * discarded along with anything after it.
 */
export function recover(log: TransitionLog, initial: State = INITIAL_STATE): RecoveryResult {
  const entries = log.read();
  let state: State = initial;
  let applied = 0;
  let discardedTornEntry = false;

  for (let i = 0; i < entries.length; i++) {
    const entry = entries[i];
    const expectedIndex = i + 1;
    if (!isEntryConsistent(entry, expectedIndex)) {
      // Torn / interrupted boundary: discard this entry and the remainder.
      discardedTornEntry = true;
      break;
    }
    // Replay is anchored to the entry's own `from` so it is total even if the
    // supplied `initial` differs; the first entry must originate at `initial`.
    state = entry.to;
    applied++;
  }

  return { state, appliedCount: applied, discardedTornEntry };
}
// #EXT-002-REQ-4 End

File: jaros/state/recover.py
"""Crash recovery by deterministic log replay (EXT-002 / REQ-4).

After a process or node crash, current state is reconstructed by replaying the
durable transition log in order. Replay is deterministic and total: the final
``state`` field of the last *valid* entry is the recovered current state.

Validation during replay:

- **Index continuity** — entries must be numbered ``1, 2, 3, ...`` with no gap.
- **Checksum** — each entry's stored checksum must match its payload.

A torn/corrupt *trailing* entry (a partially-written final append, or a final
entry that fails checksum / breaks index continuity) is discarded, yielding the
consistent pre-crash state. Corruption *before* the trailing entry indicates a
non-recoverable log and raises :class:`RecoveryError`.
"""

from __future__ import annotations

from jaros.state.log import LogEntry, TransitionLog
from jaros.state.machine import assert_valid_state
from jaros.state.model import INITIAL_STATE, STATES


# #EXT-002-REQ-4 Start
class RecoveryError(ValueError):
    """Raised when the log is corrupt in a way replay cannot safely recover."""


def recover(log: TransitionLog) -> str:
    """Replay ``log`` to reconstruct the current state.

    Reads entries in order (``read`` already drops a torn trailing *line*),
    validates index continuity and per-entry checksums, discards a single
    corrupt/torn trailing entry, and returns the resulting state. An empty (or
    fully-discarded) log recovers to :data:`INITIAL_STATE`.

    Returns:
        str: the recovered current state.

    Raises:
        RecoveryError: if corruption is detected before the trailing entry
            (i.e. the log is not safely recoverable).
    """
    entries: list[LogEntry] = list(log.read())

    valid: list[LogEntry] = []
    expected_index = 1
    for pos, entry in enumerate(entries):
        is_last = pos == len(entries) - 1
        index_ok = entry.index == expected_index
        checksum_ok = entry.checksum_ok()
        state_ok = isinstance(entry.state, str) and entry.state in STATES

        if index_ok and checksum_ok and state_ok:
            valid.append(entry)
            expected_index += 1
            continue

        # Corruption. Tolerated only if this is the trailing entry (a torn /
        # interrupted final commit); otherwise the log is unrecoverable.
        if is_last:
            break
        raise RecoveryError(
            f"corrupt log entry at position {pos} (index={entry.index!r}): "
            f"index_ok={index_ok} checksum_ok={checksum_ok} state_ok={state_ok}"
        )

    state = valid[-1].state if valid else INITIAL_STATE
    # Final invariant: recovered state must be a declared state.
    return assert_valid_state(state)
# #EXT-002-REQ-4 End
